src/ns-support.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "ns-support.h"
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sched.h>
+#include <signal.h>
+#include <string.h>
+#include <sys/eventfd.h>
+#include <sys/file.h>
+#include <sys/mount.h>
+#include <sys/prctl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#ifdef HAVE_APPARMOR
+#include <sys/apparmor.h>
+#endif				// ifdef HAVE_APPARMOR
+
+#include "utils.h"
+#include "user-support.h"
+#include "mountinfo.h"
+#include "cleanup-funcs.h"
+
+/**
+ * Directory where snap-confine keeps namespace files.
+ **/
+#define SC_NS_DIR "/run/snapd/ns"
+
+/**
+ * Effective value of SC_NS_DIR.
+ *
+ * This is only altered for testing.
+ **/
+static const char *sc_ns_dir = SC_NS_DIR;
+
+/**
+ * Name of the lock file associated with SC_NS_DIR.
+ * and a given group identifier (typically SNAP_NAME).
+ **/
+#define SC_NS_LOCK_FILE ".lock"
+
+/**
+ * Name of the preserved mount namespace associated with SC_NS_DIR
+ * and a given group identifier (typically SNAP_NAME).
+ **/
+#define SC_NS_MNT_FILE ".mnt"
+
+// Read /proc/self/mountinfo and check if /run/snapd/ns is a private bind mount.
+//
+// That is, it cannot be shared with any other peer as defined by kernel
+// documentation listed here:
+// https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt
+static bool sc_is_ns_group_dir_private();
+
+void sc_initialize_ns_groups()
+{
+	int dir_fd __attribute__ ((cleanup(sc_cleanup_close))) = -1;
+	debug("creating namespace group directory %s", sc_ns_dir);
+	mkpath(sc_ns_dir);
+	debug("opening namespace group directory %s", sc_ns_dir);
+	dir_fd = open(sc_ns_dir, O_DIRECTORY | O_PATH | O_CLOEXEC | O_NOFOLLOW);
+	if (dir_fd == -1) {
+		die("cannot open namespace group directory");
+	}
+	debug("opening lock file for group directory");
+	int lock_fd __attribute__ ((cleanup(sc_cleanup_close))) = -1;
+	lock_fd = openat(dir_fd,
+			 SC_NS_LOCK_FILE,
+			 O_CREAT | O_RDWR | O_CLOEXEC | O_NOFOLLOW, 0600);
```

Comment:
On 09/06/2016 04:17 PM, jdstrand wrote:
> In src/ns-support.c
> <https://github.com/snapcore/snap-confine/pull/126#discussion_r77721186>:
> 
>> +
>> +void sc_initialize_ns_groups()
>> +{
>> +	int dir_fd __attribute__ ((cleanup(sc_cleanup_close))) = -1;
>> +	debug("creating namespace group directory %s", sc_ns_dir);
>> +	mkpath(sc_ns_dir);
>> +	debug("opening namespace group directory %s", sc_ns_dir);
>> +	dir_fd = open(sc_ns_dir, O_DIRECTORY | O_PATH | O_CLOEXEC | O_NOFOLLOW);
>> +	if (dir_fd == -1) {
>> +		die("cannot open namespace group directory");
>> +	}
>> +	debug("opening lock file for group directory");
>> +	int lock_fd __attribute__ ((cleanup(sc_cleanup_close))) = -1;
>> +	lock_fd = openat(dir_fd,
>> +			 SC_NS_LOCK_FILE,
>> +			 O_CREAT | O_RDWR | O_CLOEXEC | O_NOFOLLOW, 0600);
> 
> Please also add O_EXCL to this openat() call (at which point, you don't
> need to specify O_NOFOLLOW).

I don't think that a lockfile should be opened with O_EXCL. Requiring
O_EXCL defeats the purpose of the lockfile since the exclusivity
provided by the lockfile is enforced by flock() once you've opened the file.

(Note that I have not reviewed the code in this PR at all but saw this
review comment fly by and thought I'd comment on it)